src/common/tusb_fifo.c

```diff
@@ -462,33 +469,74 @@ static uint16_t _tu_fifo_write_n(tu_fifo_t* f, const void * data, uint16_t n, tu
 
   _ff_lock(f->mutex_wr);
 
-  uint16_t w = f->wr_idx, r = f->rd_idx;
+  uint16_t wr_idx = f->wr_idx;
+  uint16_t rd_idx = f->rd_idx;
+
   uint8_t const* buf8 = (uint8_t const*) data;
+  uint16_t overflowable_count = _tu_fifo_count(f, wr_idx, rd_idx);
+  uint16_t const remain = _tu_fifo_remaining(f, wr_idx, rd_idx);
+
+  TU_LOG(TU_FIFO_DBG, "rd = %3u, wr = %3u, count = %3u, remain = %3u, n = %3u:  ", rd_idx, wr_idx, _tu_fifo_count(f, wr_idx, rd_idx), remain, n);
 
-  if (!f->overwritable)
+  if ( !f->overwritable )
   {
-    // Not overwritable limit up to full
-    n = tu_min16(n, _tu_fifo_remaining(f, w, r));
+    // limit up to full
+    n = tu_min16(n, remain);
   }
-  else if (n >= f->depth)
+  else
   {
-    // Only copy last part
-    buf8 = buf8 + (n - f->depth) * f->item_size;
-    n = f->depth;
-
-    // We start writing at the read pointer's position since we fill the complete
-    // buffer and we do not want to modify the read pointer within a write function!
-    // This would end up in a race condition with read functions!
-    w = r;
+    // In over-writable mode, fifo_write() is allowed even when fifo is full. In such case,
+    // oldest data in fifo i.e at read pointer data will be overwritten
+    // Note: we can modify read buffer contents but we must not modify the read index itself within a write function!
+    // Since it would end up in a race condition with read functions!
+    if ( n >= f->depth )
+    {
+      // Only copy last part
+      if ( copy_mode == TU_FIFO_COPY_INC )
+      {
+        buf8 += (n - f->depth) * f->item_size;
+      }else
+      {
+        // TODO should read from hw fifo to discard data, however reading an odd number could
+        // accidentally discard data.
+      }
+
+      n = f->depth;
+
+      // We start writing at the read pointer's position since we fill the whole buffer
+      wr_idx = rd_idx;
+    }
+    else if (overflowable_count + n >= 2*f->depth)
```

Comment:
double overflowed detection and handling with overwritable fifo

Reply:
They are overlapped a bit but not the same. I did put a note on the header file (did mentioned you there). Overflowed (single/double) only occurred with over-writable fifo where we repeatedly write() without doing any read(). Which is a case of cdc when first enumerated but dtr/rts is not set. Or with DMA but the writer is running much faster than reader. 

There is also a unite test case for this https://github.com/hathach/tinyusb/pull/1789/files/04a5c03ea8ea70510c332f0fafb44daf0ec7c7fb#diff-f49c2fc269132b486ec507c3e772fb981b9fddf2a1961ae5b9ad6d566d8d3c23R141

Reply:
I got it, maybe a comment for later: Overflow is bigger than the allowed range [0,2*depth) and thus not detect and as such not recoverable in the read function - we need to handle the overflow condition here